### PR TITLE
Simplify scratch local calculation

### PR DESCRIPTION
**Pull request by tlively**
Change `countScratchLocals` to return the count and type of necessary scratch
locals. It used to encode them as keys in the global map from scratch local
types to local indices, which could not handle having more than one scratch
local of a given type and was generally harder to reason about due to its use of
global state. Take the opportunity to avoid emitting unnecessary scratch locals
for `TupleExtract` expressions that will be optimized to not use them.

Also simplify and better document the calculation of the mapping from IR indices
to binary indices for all locals, scratch and non-scratch.

### CODE REVIEW

**Comment by tlively** [comment]
* **#6579** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6579?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6583** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6583?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @tlively and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/wasm/wasm-stack.cpp [hunk 1]
```diff
@@ -2661,44 +2672,51 @@ void BinaryInstWriter::mapLocalsAndEmitHeader() {
...
 
-void BinaryInstWriter::setScratchLocals() {
-  Index index = func->getVarIndexBase();
-  for (auto& localType : localTypes) {
-    index += numLocalsByType[localType];
-    if (scratchLocals.find(localType) != scratchLocals.end()) {
-      scratchLocals[localType] = index - 1;
+    ScratchLocalFinder(BinaryInstWriter& parent) : parent(parent) {}
+
+    // We need two i32 scratch locals for reach string slice, but they can be
```
Comment (kripken) [inline src/wasm/wasm-stack.cpp:2690]:
```suggestion
    // We need two i32 scratch locals for each string slice, but they can be
```

File: test/lit/binary/dwarf-multivalue.test [hunk 2]
```diff
@@ -97,9 +97,8 @@
 ;; ROUNDTRIP-NEXT:  (local $11 i32)
 ;; ROUNDTRIP-NEXT:  (local $12 f32)
 ;; ROUNDTRIP-NEXT:  (local $13 i32)
-;; ROUNDTRIP-NEXT:  (local $14 f32)
-;; ROUNDTRIP-NEXT:  (local $15 (tuple i32 f32))
-;; ROUNDTRIP-NEXT:  (local $16 i32)
+;; ROUNDTRIP-NEXT:  (local $14 (tuple i32 f32))
+;; ROUNDTRIP-NEXT:  (local $15 i32)
```
Comment (kripken) [inline test/lit/binary/dwarf-multivalue.test:101]: The comment on line 80 looks like it needs to be updated.

File: src/wasm/wasm-stack.cpp [hunk 3]
```diff
@@ -2661,44 +2672,51 @@ void BinaryInstWriter::mapLocalsAndEmitHeader() {
...
+    // reused.
+    bool hasStringSlice = false;
```
Comment (kripken) [inline src/wasm/wasm-stack.cpp:2692]: Is this used anywhere?

Comment (tlively) [reply src/wasm/wasm-stack.cpp:2692]: https://github.com/emscripten-core/emscripten/blob/main/src/library_wasmfs_opfs.js#L243-L278

Comment (tlively) [reply src/wasm/wasm-stack.cpp:2692]: Nope, this is left over development cruft, will remove.

File: test/lit/binary/dwarf-multivalue.test (hunk 2 above)
Comment (tlively) [reply test/lit/binary/dwarf-multivalue.test:99]: Yes, good point, I think we no longer emit `(local $14)`. I updated the comments.

**Review by aheejin — APPROVED** [review]

**Review by kripken — APPROVED** [review]

**Pull request merged**